File: LRP/Lrp.Core.Cpp/Translator.cpp
#include "stdafx.h"
#include "Translator.h"
namespace
{
	const string cWrongExceptionsSpecification = "Remote exceptions specification can not be matched to local exceptions specification.";
}

Translator::Translator() : m_componentId()
{
}
Translator::Translator(const string& text)
{
	Construct(numeric_limits<uint16>::max(), text);
}
Translator::Translator(uint16 componentId, const string& text)
{
	Construct(componentId, text);
}
const string& Translator::Name() const
{
	return m_name;
}
void Translator::Translate(uint16& componentId, uint16& methodId) const
{
	if (numeric_limits<uint16>::max() == m_componentId)
	{
		throw runtime_error("Incorrect component ID");
	}
	uint16 newMethodId = m_methods.at(methodId);
	if (numeric_limits<uint16>::max() == newMethodId)
	{
		throw runtime_error("Incorrect method ID");
	}
	componentId = m_componentId;
	methodId = newMethodId;
}
void Translator::Construct(uint16 componentId, const string& text)
{
	m_componentId = componentId;
	stringstream stream;
	stream<<text;

	if (!getline(stream, m_name, ';'))
	{
		throw runtime_error("Couldn't get component name");
	}

	string st;
	while(getline(stream, st, ';'))
	{
		m_names.push_back(st);
	}
}
void Translator::MatchMethods(const Translator* translator)
{
	if (nullptr != translator)
	{
		DoMatchMethods(*translator);
	}
	else
	{
		DoResetMethods();
	}
}
void Translator::MatchExceptions(const Translator& translator)
{
	if (translator.m_names.size() > m_names.size())
	{
		throw runtime_error(cWrongExceptionsSpecification);
	}
	size_t count = translator.m_names.size();
	for (size_t index = 0; index < count; ++index)
	{
		if (m_names[index] != translator.m_names[index])
		{
			throw runtime_error(cWrongExceptionsSpecification);
		}
	}
}
void Translator::DoMatchMethods(const Translator& translator)
{
	map<string, uint16> dict;
	for each (const auto& element in translator.m_names)
	{
		dict[element] = static_cast<uint16>(dict.size());
	}
	size_t count = this->m_names.size();
	this->m_methods.clear();

	for (size_t index = 0; index < count; ++index)
	{
		uint16 methodId = numeric_limits<uint16>::max();
		auto it = dict.find(m_names[index]);
		if (dict.end() != it)
		{
			methodId = it->second;
		}
		this->m_methods.push_back(methodId);
	}
	m_componentId = translator.m_componentId;
}
void Translator::DoResetMethods()
{
	this->m_componentId = numeric_limits<uint16>::max();
	this->m_methods.clear();
}

bool Translator::IsSupported(uint16 methodId) const
{
	if (numeric_limits<uint16>::max() == m_componentId)
	{
		return false;
	}
	if (methodId >= m_methods.size())
	{
		return false;
	}
	bool result = (numeric_limits<uint16>::max() != m_methods[methodId]);
	return result;
}
bool Translator::IsSupported() const
{
	return (numeric_limits<uint16>::max() != m_componentId);
}

void Translator::FlushMethod(CLogger& logger) const
{
	if (IsSupported())
	{
		logger.Output("{0} component is supported", m_name);
		size_t count = m_names.size();
		for (size_t index = 0; index < count; ++index)
		{
			string name = m_names[index];
			size_t position = name.find_first_of('@');
			name = name.substr(0, position);;
			if (IsSupported(static_cast<uint16>(index)))
			{
				logger.Output("\t{0} method is supported", name);
			}
			else
			{
				logger.Output("\t{0} method is not supported", name);
			}
		}
	}
	else
	{
		logger.Output("{0} component is not supported", m_name);
	}
}

File: LRP/Lrp.Core.Net/Implementation/Translator.cs
namespace SoftFX.Lrp.Implementation
{
    using System;
    using System.Collections.Generic;

    class Translator
    {
        public Translator(string st)
            : this(ushort.MaxValue, st)
        {
        }

        public Translator(ushort componentId, string st)
        {
            this.componentId = componentId;
            var separators = new char[] { ';' };
            var entries = st.Split(separators, StringSplitOptions.RemoveEmptyEntries);
            this.Name = entries[0];
            this.Methods = new string[entries.Length - 1];
            this.methodNames = new string[entries.Length - 1];

            for (var index = 1; index < entries.Length; ++index)
            {
                var method = entries[index];
                this.Methods[index - 1] = entries[index];
                var count = method.IndexOf('@');
                var name = method.Substring(0, count);
                this.methodNames[index - 1] = name;
            }
        }

        public void Translate(ref ushort componentId, ref ushort methodId)
        {
            if (this.componentId == ushort.MaxValue)
            {
                var componentName = this.Name;
                throw new InvalidComponentException(componentName, componentId);
            }

            var newMethodId = this.methods[methodId];
            if (newMethodId == ushort.MaxValue)
            {
                var componentName = this.Name;
                var methodName = this.methodNames[methodId];
                throw new InvalidMethodException(componentName, methodName, componentId, methodId);
            }

            componentId = this.componentId;
            methodId = newMethodId;
        }

        /// <summary>
        /// The method checks, if the corresponded component and method are supported by server.
        /// You should not use the method directly.
        /// </summary>
        /// <param name="methodId">a method ID</param>
        /// <returns>true, if the component and method are supported by server, otherwise false</returns>
        public bool IsSupported(ushort methodId)
        {
            if (this.componentId == ushort.MaxValue)
                return false;

            if (methodId >= this.methods.Length)
                return false;

            return this.methods[methodId] != ushort.MaxValue;
        }

        /// <summary>
        /// The method checks, if the corresponded component is supported by server.
        /// You should not use the method directly.
        /// </summary>
        /// <returns>true, if the component is supported by server, otherwise false</returns>
        public bool IsSupported()
        {
            return this.componentId != ushort.MaxValue;
        }

        public void MatchMethods(Translator translator)
        {
            if (translator != null)
                this.DoMatchMethods(translator);
            else
                this.DoResetMethods();
        }

        void DoMatchMethods(Translator translator)
        {
            var dict = new Dictionary<string, ushort>();
            foreach (var element in translator.Methods)
            {
                dict[element] = (ushort)dict.Count;
            }

            var count = this.Methods.Length;
            this.methods = new ushort[count];

            for (var index = 0; index < count; ++index)
            {
                ushort methodId;
                if (!dict.TryGetValue(this.Methods[index], out methodId))
                {
                    methodId = ushort.MaxValue;
                }
                this.methods[index] = methodId;
            }
            this.componentId = translator.componentId;
        }

        void DoResetMethods()
        {
            this.componentId = ushort.MaxValue;
            this.methods = null;
        }

        public void MatchExceptions(Translator translator)
        {
            if (translator.Methods.Length > Methods.Length)
                throw new Exception(WrongExceptionsSpecification);

            var count = translator.Methods.Length;
            for (var index = 0; index < count; ++index)
            {
                if (Methods[index] != translator.Methods[index])
                    throw new Exception(WrongExceptionsSpecification);
            }
        }

        public void FlushMethod(Logger logger)
        {
            if (this.IsSupported())
            {
                logger.Output("{0} component is supported", this.Name);
                var count = this.Methods.Length;
                for (var index = 0; index < count; ++index)
                {
                    var name = this.Methods[index];
                    var position = name.IndexOf('@');
                    name = name.Substring(0, position);
                    if (this.IsSupported((ushort)index))
                    {
                        logger.Output("\t{0} method is supported", name);
                    }
                    else
                    {
                        logger.Output("\t{0} method is not supported", name);
                    }
                }
            }
            else
            {
                logger.Output("{0} component is not supported", this.Name);
            }
        }

        public string Name { get; private set; }
        public string[] Methods { get; private set; }

        public override string ToString()
        {
            return this.Name;
        }

        #region Fields

        ushort componentId;
        ushort[] methods;
        string[] methodNames;

        #endregion

        #region Constants

        const string WrongExceptionsSpecification = "Remote exceptions specification can not be matched to local exceptions specification.";

        #endregion
    }
}
